graph/pull_v2.go

```diff
@@ -244,22 +244,32 @@ func (p *v2Puller) pullV2Tag(tag, taggedName string) (bool, error) {
 		go p.download(&downloads[i])
 	}
 
+	// Register the images serially. We must wait for all downloads to complete if one fails,
+	// in order to cleanup the others.
+	var firstErr error
 	var tagUpdated bool
 	for i := len(downloads) - 1; i >= 0; i-- {
-		d := &downloads[i]
-		if d.err != nil {
-			if err := <-d.err; err != nil {
-				return false, err
+		registerFunc := func(d *downloadInfo, pullAborted bool) error {
```

Comment:
Can you move this function definition out of the loop? It doesn't close over any variables and makes the logic harder to follow